### PR TITLE
[ASAA-178] Connecting API Calls to UI

### PR DESCRIPTION
**Pull request by MMurrayAtBR**
### [[ASAA-178] Connecting API Calls to UI](https://bottlerocketstudios.atlassian.net/browse/ASAA-178)

- Added repositoryModule to appModule
- Added missing value to DailyDataPointValuesDto
- Updated DailyForecastChip to display precipitation percentage only if the percentage is over 0
- Add `= null` to data class values where necessary
- Updated ForecastState with additional fields needed for UI elements
- Updated HomeBottomSheet to display error text field if forecasts are missing
- Updated HomeCurrentWeather to display API data correctly
- Updated HomeState (and converter) with values needed for WeatherDetails widgets
- Moved multiple files from Domain module to Data module
- Updated multiple Composable functions to check nullable values
- Updated Preview functions and MockData with new data types
- Updated ktorClient to match Ignite
- Updated NetworkRoutes with URLs and constants
- Updated TomorrowIoService with calls to get daily, hourly, and real time forecasts
- Updated UVIndexEnum and WindDirection with functions to get correct enum value
- Updated WeatherDetails for nullable values
- Removed WeatherDetailsState because it is no longer needed
- Added API calls to WeatherRepository and WeatherRepoImplementation
- Added ApiException class
- Created HomeViewModelUtils to hold all of the utility functions for HomeViewModel
- Added Domain classes, DTO classes, and converters for RealTime forecasts
- Removed TomorrowService interface on advice from Colin

**Updates for PR suggestions**
- Updated all converters to expressions
- Updated Composable functions to use `State` values
- Updated HomeState values to State 
- Removed `.stateIn()` methods from HomeViewModel
- Updated HomeViewModelUtils functions to expressions
- Updated KtorClient.kt with API Key parameter for `defaultRequest`
- Updated MockData with state values
- Removed functions from NetworkRoutes
- Added Wind Directions to String
- Updated TomorrowIoService functions with query parameters

### CODE REVIEW

File: compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/forecasts/DailyForecastChip.kt [hunk 1]
```diff
@@ -34,14 +34,16 @@ fun DailyForecastChip(state: ForecastState) {
                     contentDescription = stringResource(SharedRes.strings.weatherIcon),
                 )
                 Text(
-                    precipProbability ?: stringResource(SharedRes.strings.percentageError),
+                    stringResource(SharedRes.strings.input_percentage, precipProbability ?: "--"),
```
Comment (VanyaHuamanBottleRocket) [inline compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/forecasts/DailyForecastChip.kt:37]: could make "--" into a const

File: compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/forecasts/HourlyForecastChip.kt [hunk 2]
```diff
@@ -34,14 +34,16 @@ fun HourlyForecastChip(state: ForecastState) {
                     contentDescription = stringResource(SharedRes.strings.weatherIcon),
                 )
                 Text(
-                    precipProbability ?: stringResource(SharedRes.strings.percentageError),
+                    stringResource(SharedRes.strings.input_percentage, precipProbability ?: "0.0"),
```
Comment (VanyaHuamanBottleRocket) [inline compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/forecasts/HourlyForecastChip.kt:37]: could make "0.0" into a const

**Review by VanyaHuamanBottleRocket — APPROVED** [review]

File: compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/uvIndex/UVIndexEnum.kt [hunk 3]
```diff
@@ -16,9 +16,9 @@ enum class UVIndexEnum(
     UNKNOWN(0f, SharedRes.strings.uv_index_unknown, SharedRes.strings.unknown_uv)
 }
 
-fun Double.getUvIndex() =
-    when (this) {
-        in Double.MIN_VALUE..2.9 -> UVIndexEnum.LOW
+fun Double?.getUvIndex() =
+    when (this ?: -1.0) {
```
[the remark this answers is not in the document]
Comment (BottleRocket-Colin) [reply compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/uvIndex/UVIndexEnum.kt:20]: or use Double.MIN

File: compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/wind/WindDirectionEnum.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package com.br.kmpdemo.compose.ui.weatherDetails.wind
+
+/**
+ * "degrees" slightly adjusted from normal to better match up with
+ * Cardinal and Intermediate Lines.
+ */
+enum class WindDirectionEnum(val direction: String, val degrees: Double) {
```
[the remark this answers is not in the document]
Comment (BottleRocket-Colin) [reply compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/wind/WindDirectionEnum.kt:7]: Good idea Vanya.  the direction could be the type of the reference to the string,

File: compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/wind/WindDirectionEnum.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    WSW("West-Southwest", 240.5),
+    W("West", 270.0),
+    WNW("West-Northwest", 299.5),
+    NW("Northwest", 316.0),
+    NNW("North-Northwest", 330.5),
+    UNKNOWN("", 0.0)
+}
+
+fun Double?.getWindDirection(): WindDirectionEnum {
+    return when (this ?: -1.0) {
```
Comment (BottleRocket-Colin) [inline compose/src/commonMain/kotlin/com/br/kmpdemo/compose/ui/weatherDetails/wind/WindDirectionEnum.kt:28]: Same comment as before:  -1.0 could be Double.MIN

File: data/src/commonMain/kotlin/com/br/kmpdemo/converters/RealTimeDataDtoToRealTimeDataConverter.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.br.kmpdemo.converters
+
+import com.br.kmpdemo.dtos.RealTimeDataDto
+import com.br.kmpdemo.models.RealTimeData
+
+fun RealTimeDataDto?.toRealTimeDataModel(): RealTimeData {
+    return RealTimeData(
```
Comment (BottleRocket-Colin) [inline data/src/commonMain/kotlin/com/br/kmpdemo/converters/RealTimeDataDtoToRealTimeDataConverter.kt:7]: Could expression function

File: data/src/commonMain/kotlin/com/br/kmpdemo/converters/RealTimeDtoToRealTimeConverter.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.br.kmpdemo.converters
+
+import com.br.kmpdemo.dtos.RealTimeDto
+import com.br.kmpdemo.models.RealTime
+
+fun RealTimeDto?.toRealTimeModel(): RealTime {
```
Comment (BottleRocket-Colin) [inline data/src/commonMain/kotlin/com/br/kmpdemo/converters/RealTimeDtoToRealTimeConverter.kt:6]: expression function

File: data/src/commonMain/kotlin/com/br/kmpdemo/network/KtorClient.kt [hunk 8]
```diff
@@ -19,13 +23,28 @@ fun ktorClient(): HttpClient = HttpClient {
...
         url {
             protocol = URLProtocol.HTTPS
-            host = NetworkRoutes.BASE_HOST
-            headers {
-                appendIfNameAbsent("apikey", NetworkRoutes.API_KEY)
-            }
+//            host = NetworkRoutes.BASE_HOST
+//            headers {
+//                appendIfNameAbsent("apikey", NetworkRoutes.API_KEY)
+//            }
```
Comment (BottleRocket-Colin) [inline data/src/commonMain/kotlin/com/br/kmpdemo/network/KtorClient.kt:47]: Go ahead and delete this commented code.  Header for sure.

File: data/src/commonMain/kotlin/com/br/kmpdemo/network/service/TomorrowIoService.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            }.body<ForecastDto>()
+        }
+
+    suspend fun getDailyForecast(
+        location: String,
+        units: String,
+    ): Result<ForecastDto> =
+        runCatching {
+            client.get {
+                url(NetworkRoutes.getDailyForecast(location, units))
```
Comment (BottleRocket-Colin) [inline data/src/commonMain/kotlin/com/br/kmpdemo/network/service/TomorrowIoService.kt:31]: Let's have a chat.  We may be able to make this more KTOR idiomatic

File: shared/src/commonMain/kotlin/com/br/kmpdemo/viewmodels/HomeViewModel.kt [hunk 10]
```diff
@@ -1,55 +1,161 @@
...
+    private val dailyResponse = MutableStateFlow<Forecast?>(null)
+    val realTimeResponse = MutableStateFlow<RealTime?>(null)
+
+    /** UI Forecasts */
+    val hourlyForecasts = hourlyResponse
+        .map { it?.timelines?.hourly?.toHourlyForecastState() ?: initForecasts }
+        .stateIn(viewModelScope, SharingStarted.Lazily, null)
+    val dailyForecasts = dailyResponse
+        .map { it?.timelines?.daily?.toDailyForecastState() ?: initForecasts }
+        .stateIn(viewModelScope, SharingStarted.Lazily, null)
```
Comment (BottleRocket-Colin) [inline shared/src/commonMain/kotlin/com/br/kmpdemo/viewmodels/HomeViewModel.kt:52]: Could either create an extension function for the stateIn call, or handle default state in the converter function.

File: shared/src/commonMain/kotlin/com/br/kmpdemo/viewmodels/HomeViewModelUtils.kt [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        return normalizedValue.coerceIn(0.0F, 1.0F)
+    }
+
+    /// Use the next to functions to find the forecast for today or this hour
+    fun String.isToday(): Boolean {
+        val today = Clock.System.now()
+            .toLocalDateTime(TimeZone.currentSystemDefault())
+            .toString()
+            .toReformattedDate()
+        return this.toReformattedDate() == today
```
Comment (BottleRocket-Colin) [inline shared/src/commonMain/kotlin/com/br/kmpdemo/viewmodels/HomeViewModelUtils.kt:27]: Some small tinkers here could make this an expression.

**Review by BottleRocket-Colin — CHANGES_REQUESTED** [review]
really awesome PR.  Some minor changes to compress code.

**Pull request merged**